jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/PerMessageDeflateExtension.java

```diff
@@ -335,85 +354,131 @@ else if (_frame.isFin())
         }
     }
 
-    private class IncomingFlusher extends TransformingFlusher
+    private class IncomingFlusher extends IteratingCallback
     {
+        private final AtomicLong _demand = new AtomicLong();
+        private final AtomicReference<Throwable> _failure = new AtomicReference<>();
+        private boolean _finished = true;
         private boolean _first;
         private Frame _frame;
+        private ByteBuffer _framePayload;
+        private Callback _frameCallback;
         private boolean _tailBytes;
+        private LongConsumer _nextDemand;
 
-        @Override
-        protected boolean onFrame(Frame frame, Callback callback, boolean batch)
+        public boolean isFinished()
         {
-            _tailBytes = false;
-            _first = true;
-            _frame = frame;
+            return _finished;
+        }
+
+        public void setNextDemand(LongConsumer nextDemand)
+        {
+            _nextDemand = nextDemand;
+        }
+
+        public void demand(long n)
+        {
+            if (n <= 0)
+                throw new IllegalArgumentException("Demand must be positive");
+            _demand.addAndGet(n);
```

Comment:
Cap the demand with `MathUtils`.

Reply:
Ok, but will use `Math.addExact` instead as its probably more what we want.

You don't really want to cap the demand because then its out of sync with what the application is expecting, although this is not likely to be a problem with `long`.